File: leetcode216.py
class Solution:
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """

        def helper(i,k,n,res,seq):
            if i == 10:
                if k == 0 and n == 0:
                    res.append(seq)
                return 

            if n >= i and k > 0:
                helper(i+1,k-1,n-i,res,seq+[i])

            helper(i+1,k,n,res,seq)

        res = []
        helper(1,k,n,res,[])
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.combinationSum3(3,7))

File: leetcodec++/leetcode216.cpp
//
// Created by 黄敏辉 on 2019-09-02.
//

class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> comb;
        find_comb(res,comb, k, n);
        return res;
    }

    void find_comb(vector<vector<int>>& res, vector<int>& comb, int k ,int n){
        if(comb.size() == k && n == 0){
            res.push_back(comb);
            return;
        }
        if(comb.size() < k){
            for (int i = comb.empty()? 1:comb.back()+1; i <= 9;i++){
                if (n - i < 0)
                    break;
                comb.push_back(i);
                find_comb(res, comb, k, n-i);
                comb.pop_back();
            }
        }
    }
};